### PR TITLE
Fix/insert bug

### LINKED ISSUE #121: Cannot import API
Cannot import the stub APi at /examples/apis/posts.json. It fails with this response:

HTTP/1.1 400 Bad Request
Content-Length: 27
Content-Type: application/json
Date: Wed, 05 Apr 2017 22:43:03 GMT
Vary: Accept-Encoding

"api definition not found"

**Reproduction Steps:**

1.  Start all docker containers with: docker-compose up -d 
2.  Create valid JWT token
3.  Submit this request from within the janus-master folder via CLI:

http POST 192.168.99.100:8081/apis "Authorization:Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE0OTE0MzM3MTIsImlhdCI6MTQ5MTQzMDExMiwiaWQiOiJhZG1pbiJ9.CfNK3mxh3yArgrDXBLpYArLqPU6PywIDCxBdNEeSQUA" "Content-Type: application/json" < examples/apis/posts.json

**Expected behavior:**

Successful addition of jsonplaceholder API

**Observed behavior:**

API definition not found error.

**Janus version: latest version
**OS and version: Mac OSX 10.11.6

### PR DESCRIPTION
**Pull request by italolelis**
## What does this PR do?

- Fixes the bug for inserting a new API

fixes #121

### CODE REVIEW

File: pkg/api/error.go [hunk 1]
```diff
@@ -7,6 +7,12 @@ var (
 	// ErrAPIDefinitionNotFound is used when the api was not found in the datastore
 	ErrAPIDefinitionNotFound = errors.New(http.StatusNotFound, "api definition not found")
 
+	// ErrAPINameExists is used when the API name is already registeres on the datastore
+	ErrAPINameExists = errors.New(http.StatusBadRequest, "api name is already registered")
+
+	// ErrAPIListenPathExists is used when the API name is already registeres on the datastore
```
Comment (vgarvardt) [inline pkg/api/error.go]: ErrAPIListenPathExists is used when the API ~~name~~ listen path is already registeres on the datastore

File: pkg/api/mongodb_repository.go [hunk 2]
```diff
@@ -53,15 +53,30 @@ func (r *MongoRepository) FindByName(name string) (*Definition, error) {
...
 	session, coll := r.getSession()
 	defer session.Close()
 
-	err := coll.Find(bson.M{"proxy.listen_path": path}).One(&result)
+	count, err := coll.Find(bson.M{"name": def.Name}).Count()
+	if err != nil {
+		return false, err
+	}
+
+	if count <= 0 {
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go]: `count >= 1` ?

File: pkg/api/mongodb_repository.go [hunk 3]
```diff
@@ -53,15 +53,30 @@ func (r *MongoRepository) FindByName(name string) (*Definition, error) {
...
+		return false, ErrAPINameExists
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go]: `return true, ErrAPINameExists `

File: pkg/api/mongodb_repository.go [hunk 4]
```diff
@@ -53,15 +53,30 @@ func (r *MongoRepository) FindByName(name string) (*Definition, error) {
...
+	}
 
-	return result, err
+	count, err = coll.Find(bson.M{"proxy.listen_path": def.Proxy.ListenPath}).Count()
+	if err != nil {
+		return false, err
+	}
+
+	if count <= 0 {
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go]: `count >= 1` ?

File: pkg/api/mongodb_repository.go [hunk 5]
```diff
@@ -53,15 +53,30 @@ func (r *MongoRepository) FindByName(name string) (*Definition, error) {
...
+		return false, ErrAPIListenPathExists
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go]: `return true, ErrAPIListenPathExists `

File: pkg/api/mongodb_repository.go [hunk 6]
```diff
@@ -53,15 +53,30 @@ func (r *MongoRepository) FindByName(name string) (*Definition, error) {
...
+	}
+
+	return true, nil
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go]: `return false, nil`

**Review by vgarvardt — CHANGES_REQUESTED** [review]

File: pkg/api/file_repository.go [hunk 7]
```diff
@@ -79,15 +79,20 @@ func (r *FileSystemRepository) FindByName(name string) (*Definition, error) {
 	return definition, nil
 }
 
-// FindByListenPath searches an existing Proxy definition by its listen_path
-func (r *FileSystemRepository) FindByListenPath(path string) (*Definition, error) {
+// Exists searches an existing Proxy definition by its listen_path
+func (r *FileSystemRepository) Exists(def *Definition) (bool, error) {
+	_, ok := r.definitions[def.Name]
```
Comment (rafaeljesus) [inline pkg/api/file_repository.go]: Reading from maps is not thread safe, we should use `sync.RWMutex` `RLock()` method

**Review by rafaeljesus — CHANGES_REQUESTED** [review]
Reading from maps is not thread safe per default

**Review by rafaeljesus — APPROVED** [review]

**Comment by rafaeljesus** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/128/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Review by vgarvardt — APPROVED** [review]

**Comment by vgarvardt** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/128/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Pull request merged**